{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e9d3d09",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import mygene\n",
    "import collections\n",
    "import math\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "248c75bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pip\n",
    "# pip.main(['install','mygene'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ae6b6b",
   "metadata": {},
   "source": [
    "# 1. GDSC data processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5dca0f2",
   "metadata": {},
   "source": [
    "## 1.1 GDSC drug sensitivity data binarized"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f550967b",
   "metadata": {},
   "source": [
    "### 1.1.1 read GDSC drug sensitivity data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dd659ed5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gdscDrugFile = \"GDSC_original_data/GDSC1_fitted_dose_response_25Feb20.xlsx\"\n",
    "dropIndexList = [687794, 688015, 753597, 905943, 906825, 907063, 908118, 908120, 908134, 908442, 908480, \\\n",
    "             909762, 910918, 924189, 930301, 946363, 949092, 949154, 1240149, 1240157, 1240158, 1240220, \\\n",
    "             1240224, 1298144, 1298167, 1298218, 1299069, 1303901, 1303911, 1330960, 1330981, 1331030, \\\n",
    "             1480367, 1789883] # These cell lines are droped for GDSC1 drug response data due to data incompleteness\n",
    "responseOutput = \"processed_data/GDSC1_drug_response_binaryAUC.csv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "155f44ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of cell lines with drug sensitivity data: 987\n",
      "The number of drugs: 367\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>PATHWAY_NAME</th>\n",
       "      <th>COMPANY_ID</th>\n",
       "      <th>WEBRELEASE</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>MAX_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12974350</td>\n",
       "      <td>683665</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>MM</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975300</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975647</td>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975980</td>\n",
       "      <td>684059</td>\n",
       "      <td>ES7</td>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.692768</td>\n",
       "      <td>0.972699</td>\n",
       "      <td>0.110056</td>\n",
       "      <td>0.088760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12976330</td>\n",
       "      <td>684062</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.478678</td>\n",
       "      <td>0.944462</td>\n",
       "      <td>0.087011</td>\n",
       "      <td>-0.111820</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "0   GDSC1             281       12974350     683665         MC-CAR   \n",
       "1   GDSC1             281       12975300     684055            ES3   \n",
       "2   GDSC1             281       12975647     684057            ES5   \n",
       "3   GDSC1             281       12975980     684059            ES7   \n",
       "4   GDSC1             281       12976330     684062          EW-11   \n",
       "\n",
       "  SANGER_MODEL_ID     TCGA_DESC  DRUG_ID  DRUG_NAME PUTATIVE_TARGET  \\\n",
       "0       SIDM00636            MM        1  Erlotinib            EGFR   \n",
       "1       SIDM00265  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "2       SIDM00263  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "3       SIDM00269  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "4       SIDM00203  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "\n",
       "     PATHWAY_NAME  COMPANY_ID WEBRELEASE  MIN_CONC  MAX_CONC   LN_IC50  \\\n",
       "0  EGFR signaling        1045          Y  0.007813       2.0  2.395685   \n",
       "1  EGFR signaling        1045          Y  0.007813       2.0  3.140923   \n",
       "2  EGFR signaling        1045          Y  0.007813       2.0  3.968757   \n",
       "3  EGFR signaling        1045          Y  0.007813       2.0  2.692768   \n",
       "4  EGFR signaling        1045          Y  0.007813       2.0  2.478678   \n",
       "\n",
       "        AUC      RMSE   Z_SCORE  \n",
       "0  0.982114  0.022521 -0.189576  \n",
       "1  0.984816  0.031840  0.508635  \n",
       "2  0.985693  0.026052  1.284229  \n",
       "3  0.972699  0.110056  0.088760  \n",
       "4  0.944462  0.087011 -0.111820  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdscDrugData = pd.read_excel(gdscDrugFile, header = 0)\n",
    "print(\"The number of cell lines with drug sensitivity data: \" + str(len(gdscDrugData.CELL_LINE_NAME.unique())))\n",
    "print(\"The number of drugs: \" + str(len(gdscDrugData.DRUG_ID.unique())))\n",
    "gdscDrugData.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "065f88c5",
   "metadata": {},
   "source": [
    "### 1.1.2 obtain the AUC value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "510e2eba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(987, 367)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>3</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>9</th>\n",
       "      <th>11</th>\n",
       "      <th>17</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>32</th>\n",
       "      <th>...</th>\n",
       "      <th>1494</th>\n",
       "      <th>1495</th>\n",
       "      <th>1496</th>\n",
       "      <th>1497</th>\n",
       "      <th>1498</th>\n",
       "      <th>1502</th>\n",
       "      <th>1526</th>\n",
       "      <th>1527</th>\n",
       "      <th>1529</th>\n",
       "      <th>1530</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.980891</td>\n",
       "      <td>0.903979</td>\n",
       "      <td>0.986077</td>\n",
       "      <td>0.937027</td>\n",
       "      <td>0.728050</td>\n",
       "      <td>0.978113</td>\n",
       "      <td>0.945160</td>\n",
       "      <td>0.988067</td>\n",
       "      <td>0.897546</td>\n",
       "      <td>...</td>\n",
       "      <td>0.650553</td>\n",
       "      <td>0.984669</td>\n",
       "      <td>0.935993</td>\n",
       "      <td>0.987922</td>\n",
       "      <td>0.979853</td>\n",
       "      <td>0.982706</td>\n",
       "      <td>0.928560</td>\n",
       "      <td>0.859750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683667</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.685058</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.941059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684052</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.290396</td>\n",
       "      <td>0.784440</td>\n",
       "      <td>0.664747</td>\n",
       "      <td>0.983333</td>\n",
       "      <td>0.919021</td>\n",
       "      <td>0.983979</td>\n",
       "      <td>0.921035</td>\n",
       "      <td>0.655726</td>\n",
       "      <td>0.621239</td>\n",
       "      <td>0.984947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684055</th>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.981415</td>\n",
       "      <td>0.941416</td>\n",
       "      <td>0.981549</td>\n",
       "      <td>0.956082</td>\n",
       "      <td>0.975094</td>\n",
       "      <td>0.979193</td>\n",
       "      <td>0.926046</td>\n",
       "      <td>0.924635</td>\n",
       "      <td>0.976683</td>\n",
       "      <td>...</td>\n",
       "      <td>0.296371</td>\n",
       "      <td>0.803190</td>\n",
       "      <td>0.671472</td>\n",
       "      <td>0.961318</td>\n",
       "      <td>0.973103</td>\n",
       "      <td>0.974498</td>\n",
       "      <td>0.963667</td>\n",
       "      <td>0.816016</td>\n",
       "      <td>0.818002</td>\n",
       "      <td>0.983505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684057</th>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.980594</td>\n",
       "      <td>0.951725</td>\n",
       "      <td>0.961611</td>\n",
       "      <td>0.917021</td>\n",
       "      <td>0.981629</td>\n",
       "      <td>0.981101</td>\n",
       "      <td>0.934873</td>\n",
       "      <td>0.963014</td>\n",
       "      <td>0.982634</td>\n",
       "      <td>...</td>\n",
       "      <td>0.391312</td>\n",
       "      <td>0.854097</td>\n",
       "      <td>0.854424</td>\n",
       "      <td>0.950922</td>\n",
       "      <td>0.961995</td>\n",
       "      <td>0.967201</td>\n",
       "      <td>0.976020</td>\n",
       "      <td>0.596586</td>\n",
       "      <td>0.796861</td>\n",
       "      <td>0.973428</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 367 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            1         3         5         6         9         11        17    \\\n",
       "683665  0.982114  0.980891  0.903979  0.986077  0.937027  0.728050  0.978113   \n",
       "683667       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "684052       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "684055  0.984816  0.981415  0.941416  0.981549  0.956082  0.975094  0.979193   \n",
       "684057  0.985693  0.980594  0.951725  0.961611  0.917021  0.981629  0.981101   \n",
       "\n",
       "            29        30        32    ...      1494      1495      1496  \\\n",
       "683665  0.945160  0.988067  0.897546  ...  0.650553  0.984669  0.935993   \n",
       "683667       NaN       NaN       NaN  ...  0.685058       NaN  0.941059   \n",
       "684052       NaN       NaN       NaN  ...  0.290396  0.784440  0.664747   \n",
       "684055  0.926046  0.924635  0.976683  ...  0.296371  0.803190  0.671472   \n",
       "684057  0.934873  0.963014  0.982634  ...  0.391312  0.854097  0.854424   \n",
       "\n",
       "            1497      1498      1502      1526      1527      1529      1530  \n",
       "683665  0.987922  0.979853  0.982706  0.928560  0.859750       NaN       NaN  \n",
       "683667       NaN       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "684052  0.983333  0.919021  0.983979  0.921035  0.655726  0.621239  0.984947  \n",
       "684055  0.961318  0.973103  0.974498  0.963667  0.816016  0.818002  0.983505  \n",
       "684057  0.950922  0.961995  0.967201  0.976020  0.596586  0.796861  0.973428  \n",
       "\n",
       "[5 rows x 367 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index = sorted(gdscDrugData.COSMIC_ID.unique())\n",
    "columns = sorted(gdscDrugData.DRUG_ID.unique())\n",
    "tmpInd = 0\n",
    "cellidDic = {}\n",
    "for ind in index:\n",
    "    cellidDic[ind] = tmpInd\n",
    "    tmpInd += 1\n",
    "tmpInd = 0\n",
    "drugnameDic = {}\n",
    "for col in columns:\n",
    "    drugnameDic[col] = tmpInd\n",
    "    tmpInd += 1\n",
    "drugResponse = np.empty((len(index), len(columns)))\n",
    "drugResponse[:,:] = np.nan\n",
    "for ind, row in gdscDrugData.iterrows():\n",
    "    cellidInd = cellidDic[row[\"COSMIC_ID\"]]\n",
    "    drugnameInd = drugnameDic[row[\"DRUG_ID\"]]\n",
    "    auc = row[\"AUC\"]\n",
    "    drugResponse[cellidInd][drugnameInd] = auc\n",
    "drugResponseDF = pd.DataFrame(drugResponse, index = index, columns = columns)\n",
    "print(drugResponseDF.shape)\n",
    "drugResponseDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "247cb9d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(953, 367)\n"
     ]
    }
   ],
   "source": [
    "drugResponseDF = drugResponseDF.drop(dropIndexList)\n",
    "print(drugResponseDF.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39600a19",
   "metadata": {},
   "source": [
    "### 1.1.3 Binarize response value with the waterfall method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b3612092",
   "metadata": {},
   "outputs": [],
   "source": [
    "def lineFromPoints(P,Q): \n",
    "    a = Q[1] - P[1] \n",
    "    b = P[0] - Q[0]  \n",
    "    c = a*(P[0]) + b*(P[1])  \n",
    "    return a, b, c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "29a3fd4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "drugThreshold = {}\n",
    "for drug in drugResponseDF.columns:\n",
    "    resVector = drugResponseDF[drug].values\n",
    "    resVector = resVector[~np.isnan(resVector)] \n",
    "    n = resVector.shape[0]\n",
    "    y = np.sort(resVector)\n",
    "    x = np.array([i for i in range(1, n + 1)])\n",
    "    pr = stats.pearsonr(y, x)[0]\n",
    "    if pr > 0.95: # Use the median as the threshold if the response curve is close to a straight line\n",
    "        drugThreshold[drug] = np.median(y)\n",
    "    else: # Use the waterfall method for finding the threshold\n",
    "        a, b, c = lineFromPoints([x[0], y[0]], [x[-1], y[-1]])\n",
    "        constant = math.sqrt(a**2 + b**2)\n",
    "        distance = np.abs(a*x + b*y + c) / constant\n",
    "        drugThreshold[drug] = y[np.argmax(distance)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c065ddd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of valid drugs:  352\n",
      "Average total number of celllines:  821.7528409090909\n",
      "Average prevalence:  0.24553267045454544\n",
      "(953, 367)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>3</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>9</th>\n",
       "      <th>11</th>\n",
       "      <th>17</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>32</th>\n",
       "      <th>...</th>\n",
       "      <th>1494</th>\n",
       "      <th>1495</th>\n",
       "      <th>1496</th>\n",
       "      <th>1497</th>\n",
       "      <th>1498</th>\n",
       "      <th>1502</th>\n",
       "      <th>1526</th>\n",
       "      <th>1527</th>\n",
       "      <th>1529</th>\n",
       "      <th>1530</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683667</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684052</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684055</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684057</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 367 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        1     3     5     6     9     11    17    29    30    32    ...  1494  \\\n",
       "683665   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0  ...   1.0   \n",
       "683667   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   0.0   \n",
       "684052   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   1.0   \n",
       "684055   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   \n",
       "684057   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   \n",
       "\n",
       "        1495  1496  1497  1498  1502  1526  1527  1529  1530  \n",
       "683665   0.0   0.0   0.0   0.0   0.0   0.0   0.0   NaN   NaN  \n",
       "683667   NaN   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN  \n",
       "684052   1.0   1.0   0.0   0.0   0.0   0.0   1.0   1.0   0.0  \n",
       "684055   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  \n",
       "684057   1.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0  \n",
       "\n",
       "[5 rows x 367 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevalence = {}\n",
    "totalCellline = {}\n",
    "binaryDrugResponseDF = pd.DataFrame(np.nan, index = drugResponseDF.index, columns = drugResponseDF.columns)\n",
    "for drug in binaryDrugResponseDF.columns:\n",
    "    sens = drugResponseDF[drug] <= drugThreshold[drug]\n",
    "    resi = drugResponseDF[drug] > drugThreshold[drug]\n",
    "    if sum(sens) + sum(resi) > 200 and round(sum(sens) / (sum(sens) + sum(resi)), 4) >= 0.05:\n",
    "        prevalence[drug] = round(sum(sens) / (sum(sens) + sum(resi)), 4)\n",
    "        totalCellline[drug] = sum(sens) + sum(resi)\n",
    "    binaryDrugResponseDF[drug][sens] = 1\n",
    "    binaryDrugResponseDF[drug][resi] = 0\n",
    "print(\"Number of valid drugs: \", len(totalCellline))\n",
    "print(\"Average total number of celllines: \", np.mean(list(totalCellline.values())))\n",
    "print(\"Average prevalence: \", np.mean(list(prevalence.values())))\n",
    "print(binaryDrugResponseDF.shape)\n",
    "binaryDrugResponseDF.to_csv(responseOutput, na_rep = \"NA\")\n",
    "binaryDrugResponseDF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1c73cbf",
   "metadata": {},
   "source": [
    "### 1.2 mutation data from a paper called \"A Landscape of Pharmacogenomic Interactions in Cancer \". "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9f0638bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/shr81/.conda/envs/MachineLearning/lib/python3.7/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['TableS2A-CancerGenes', 'TableS2B-TumourVariants', 'TableS2C-CellLineVariants', 'TableS2D-RACSs', 'TableS2E-RACSsPanCanVSCanSpec', 'TableS2F-TumoursRACSs_CNA', 'TableS2G-CellLinesRACSs_CNA', 'TableS2H-InformativeCpGislands', 'TableS2I-TumoursHypMet_iCgPs', 'TableS2J-CellLinesHypMet_iCgPs', 'TableS2K-missingCFEs', 'TableS2L-CFEs_correlations', 'TableS2M-MergedGlobalPWcorrels'])\n"
     ]
    }
   ],
   "source": [
    "xl_file = pd.ExcelFile(\"GDSC_original_data/mmc3.xlsx\")\n",
    "dfs = {sheet_name: xl_file.parse(sheet_name) \n",
    "          for sheet_name in xl_file.sheet_names}\n",
    "print(dfs.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "14774664",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function could transfor excel table to matrix for mutation data and cna data\n",
    "def excel_to_matrix_mutation(sheet_name,sample_start_point,sample_id_column_number,cancer_type_column_number,gene_column_number):\n",
    "    variants_original_df = pd.DataFrame(data=dfs[sheet_name].iloc[sample_start_point:,[cancer_type_column_number,gene_column_number]].values,index=dfs[sheet_name].iloc[sample_start_point:,sample_id_column_number].values,columns=[\"cancer_type\",\"gene\"])\n",
    "    unique_sample =variants_original_df.index.unique()\n",
    "    unique_gene = variants_original_df.gene.unique()\n",
    "    print(\"number of samples:\",len(unique_sample),\"number of genes:\",len(unique_gene))\n",
    "    variants_df = pd.DataFrame(data=np.zeros((len(unique_sample),len(unique_gene))),index=unique_sample,columns=unique_gene) #create an empty matrix with all zeros\n",
    "    for sample_name in variants_df.index:\n",
    "        res = variants_original_df.loc[sample_name,\"gene\"]\n",
    "        mutated_genes = set([res]) if isinstance(res, str) else set(res.unique())    \n",
    "        variants_df.loc[sample_name,mutated_genes] = 1 #get the mutation genes for each sample and set the these blocks to be 1\n",
    "    variants_original_drop_duplicate_sample_df = variants_original_df[~variants_original_df.index.duplicated(keep='first')]\n",
    "    variants_df[\"cancer_type\"]= variants_original_drop_duplicate_sample_df.loc[variants_df.index,\"cancer_type\"] #add cancer type to the dataframe\n",
    "    return variants_df,variants_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4848f152",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell line mutation information\n",
      "number of samples: 1001 number of genes: 19100\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1001, 19101)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Cell line mutation information\")\n",
    "Cellline_mutation_cancer_type_df,cellline_mutation_gene_names = excel_to_matrix_mutation('TableS2C-CellLineVariants',20,2,3,4)\n",
    "Cellline_mutation_cancer_type_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f79543bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4D226</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>907272</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998184</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907289</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905989</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909698</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910209</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1240132</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1290807</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330932</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688058</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1001 rows × 19100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         A1BG  A1CF  A2M  A2ML1  A4D226  A4GALT  A4GNT  AAAS  AACS  AADAC  \\\n",
       "907272    1.0   1.0  1.0    0.0     0.0     1.0    0.0   1.0   0.0    0.0   \n",
       "998184    1.0   0.0  1.0    0.0     1.0     0.0    0.0   0.0   0.0    0.0   \n",
       "907289    1.0   1.0  0.0    1.0     0.0     0.0    0.0   0.0   0.0    1.0   \n",
       "905989    1.0   0.0  0.0    1.0     0.0     0.0    0.0   0.0   0.0    0.0   \n",
       "909698    1.0   0.0  1.0    0.0     0.0     0.0    0.0   0.0   0.0    0.0   \n",
       "...       ...   ...  ...    ...     ...     ...    ...   ...   ...    ...   \n",
       "910209    0.0   0.0  0.0    0.0     0.0     0.0    0.0   0.0   0.0    0.0   \n",
       "1240132   0.0   0.0  0.0    0.0     0.0     0.0    0.0   0.0   0.0    0.0   \n",
       "1290807   0.0   0.0  0.0    0.0     0.0     0.0    0.0   0.0   0.0    0.0   \n",
       "1330932   0.0   0.0  0.0    0.0     0.0     0.0    0.0   0.0   0.0    0.0   \n",
       "688058    0.0   0.0  0.0    0.0     0.0     0.0    0.0   0.0   0.0    0.0   \n",
       "\n",
       "         ...  ZWILCH  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  ZYX  ZZEF1  \\\n",
       "907272   ...     0.0    0.0   0.0   0.0   0.0     0.0     0.0  0.0    1.0   \n",
       "998184   ...     0.0    0.0   0.0   0.0   1.0     0.0     0.0  1.0    1.0   \n",
       "907289   ...     1.0    0.0   0.0   0.0   0.0     1.0     0.0  0.0    1.0   \n",
       "905989   ...     0.0    0.0   0.0   0.0   0.0     0.0     0.0  0.0    1.0   \n",
       "909698   ...     0.0    0.0   0.0   1.0   0.0     0.0     0.0  0.0    0.0   \n",
       "...      ...     ...    ...   ...   ...   ...     ...     ...  ...    ...   \n",
       "910209   ...     0.0    0.0   0.0   0.0   0.0     0.0     0.0  0.0    0.0   \n",
       "1240132  ...     0.0    0.0   0.0   0.0   0.0     0.0     0.0  0.0    0.0   \n",
       "1290807  ...     0.0    0.0   0.0   0.0   0.0     0.0     0.0  0.0    0.0   \n",
       "1330932  ...     0.0    0.0   0.0   0.0   0.0     0.0     0.0  0.0    0.0   \n",
       "688058   ...     0.0    0.0   0.0   0.0   0.0     0.0     0.0  0.0    0.0   \n",
       "\n",
       "         ZZZ3  \n",
       "907272    0.0  \n",
       "998184    0.0  \n",
       "907289    1.0  \n",
       "905989    0.0  \n",
       "909698    0.0  \n",
       "...       ...  \n",
       "910209    0.0  \n",
       "1240132   0.0  \n",
       "1290807   0.0  \n",
       "1330932   0.0  \n",
       "688058    0.0  \n",
       "\n",
       "[1001 rows x 19100 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cellline_mutation_df = Cellline_mutation_cancer_type_df.iloc[:,:-1]\n",
    "Cellline_mutation_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "854edb42",
   "metadata": {},
   "source": [
    "### 1.3 CNV data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "446d0d07",
   "metadata": {},
   "outputs": [],
   "source": [
    "GDSC_cnv_df = pd.read_csv(\"GDSC_original_data/WES_pureCN_CNV_genes_cn_category_20220623.csv\",index_col=0) #cnv data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c5ddbe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "Cellline_info_raw_df = pd.read_csv(\"GDSC_original_data/model_list_20220810.csv\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4b7b381b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/shr81/.conda/envs/MachineLearning/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "# choose data source equals to sanger, remove .1 from sample name, and convert text to binarized form\n",
    "cell_model_passport_cnv_T_df = GDSC_cnv_df.T\n",
    "cell_model_passport_cnv_T_df = cell_model_passport_cnv_T_df[cell_model_passport_cnv_T_df[\"source\"]==\"Sanger\"]\n",
    "cell_model_passport_cnv_T_df.index = cell_model_passport_cnv_T_df.index.str.replace(\"\\.1\",\"\")\n",
    "cell_model_passport_cnv_T_df = cell_model_passport_cnv_T_df.iloc[:,3:] #remove column model_id, source, symbol\n",
    "cell_model_passport_cnv_T_df = cell_model_passport_cnv_T_df.replace([\"Neutral\",\"Loss\",\"Gain\",\"Amplification\",\"Deletion\"],[0,0,0,1,1]) # \"Neutral\",\"Loss\",\"Gain\" to 0,amplification and deletion to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6f47b52e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1008\n"
     ]
    }
   ],
   "source": [
    "#since mutation and expression data use cosmic_id, convert the cnv data from model name to cosmic_id\n",
    "cell_line_info_df = Cellline_info_raw_df[[\"model_name\",'COSMIC_ID']]\n",
    "cell_line_info_df = cell_line_info_df.dropna()\n",
    "cell_line_info_df = cell_line_info_df.set_index(\"model_name\")\n",
    "common_model_name_list = list(set(cell_model_passport_cnv_T_df.index)&set(cell_line_info_df.index))\n",
    "print(len(common_model_name_list))\n",
    "cell_line_info_df = cell_line_info_df.loc[common_model_name_list,:]\n",
    "cell_model_passport_cnv_T_df =cell_model_passport_cnv_T_df.loc[common_model_name_list,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b7cec69",
   "metadata": {},
   "source": [
    "### 1.4 SGA data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c3074067",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SDHB</th>\n",
       "      <th>PPP1R14D</th>\n",
       "      <th>WNK1</th>\n",
       "      <th>PEX26</th>\n",
       "      <th>ATXN7L3</th>\n",
       "      <th>VGF</th>\n",
       "      <th>PIP5K1C</th>\n",
       "      <th>GOLGA7</th>\n",
       "      <th>ANKRD50</th>\n",
       "      <th>ZNF35</th>\n",
       "      <th>...</th>\n",
       "      <th>MEP1B</th>\n",
       "      <th>CDKL1</th>\n",
       "      <th>HDGFL1</th>\n",
       "      <th>INSM2</th>\n",
       "      <th>TYMS</th>\n",
       "      <th>TRAM2</th>\n",
       "      <th>MIPEP</th>\n",
       "      <th>NAT1</th>\n",
       "      <th>CHRNA6</th>\n",
       "      <th>SLC25A15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>907268</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907269</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907270</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907271</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907272</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753624</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908135</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688087</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>917486</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688121</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>984 rows × 15817 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        SDHB  PPP1R14D  WNK1  PEX26  ATXN7L3  VGF  PIP5K1C  GOLGA7  ANKRD50  \\\n",
       "907268     0         0     0      1        0    0        0       0        0   \n",
       "907269     0         0     0      0        0    0        0       0        0   \n",
       "907270     0         0     0      0        0    0        0       0        0   \n",
       "907271     0         0     0      1        0    0        0       0        0   \n",
       "907272     0         0     0      0        1    1        1       0        0   \n",
       "...      ...       ...   ...    ...      ...  ...      ...     ...      ...   \n",
       "753624     0         0     0      0        1    0        0       0        0   \n",
       "908135     0         0     0      0        0    0        0       0        0   \n",
       "688087     0         0     0      0        0    0        0       0        0   \n",
       "917486     0         0     1      0        0    0        0       0        0   \n",
       "688121     0         0     0      0        0    0        0       0        0   \n",
       "\n",
       "        ZNF35  ...  MEP1B  CDKL1  HDGFL1  INSM2  TYMS  TRAM2  MIPEP  NAT1  \\\n",
       "907268      0  ...      0      0       0      0     0      0      0     0   \n",
       "907269      0  ...      0      0       0      0     0      0      0     0   \n",
       "907270      0  ...      0      0       0      0     0      0      0     0   \n",
       "907271      0  ...      0      0       0      0     0      0      0     0   \n",
       "907272      0  ...      1      0       0      0     0      0      0     0   \n",
       "...       ...  ...    ...    ...     ...    ...   ...    ...    ...   ...   \n",
       "753624      0  ...      0      0       0      0     0      0      0     0   \n",
       "908135      0  ...      0      0       0      0     0      0      0     0   \n",
       "688087      0  ...      0      0       0      0     0      0      0     0   \n",
       "917486      0  ...      0      0       0      0     0      0      0     0   \n",
       "688121      0  ...      0      0       0      0     0      0      0     0   \n",
       "\n",
       "        CHRNA6  SLC25A15  \n",
       "907268       0         0  \n",
       "907269       0         0  \n",
       "907270       0         0  \n",
       "907271       0         0  \n",
       "907272       0         0  \n",
       "...        ...       ...  \n",
       "753624       0         0  \n",
       "908135       0         0  \n",
       "688087       0         0  \n",
       "917486       0         0  \n",
       "688121       0         0  \n",
       "\n",
       "[984 rows x 15817 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_model_passport_cnv_T_df = cell_model_passport_cnv_T_df.set_index(cell_line_info_df[\"COSMIC_ID\"])\n",
    "cell_model_passport_cnv_T_df = cell_model_passport_cnv_T_df.rename(index={'907284;1330932':'1330932'})\n",
    "cell_model_passport_cnv_T_df = cell_model_passport_cnv_T_df.rename(index={'1331031;1331030':'1331030'})\n",
    "cell_model_passport_cnv_T_df.index = cell_model_passport_cnv_T_df.index.astype(int)\n",
    "\n",
    "common_sample_cnv_mut_list = list(set(Cellline_mutation_df.index)&set(cell_model_passport_cnv_T_df.index))\n",
    "common_gene_cnv_mut_list = list(set(Cellline_mutation_df.columns)&set(cell_model_passport_cnv_T_df.columns))\n",
    "# using common sample and gene to reconstruct these two dataframes\n",
    "Cellline_mutation_df = Cellline_mutation_df.loc[common_sample_cnv_mut_list,common_gene_cnv_mut_list]\n",
    "cell_model_passport_cnv_T_df = cell_model_passport_cnv_T_df.loc[common_sample_cnv_mut_list,common_gene_cnv_mut_list]\n",
    "#if a gene is 1 in mutation or cnv, then it is 1, neither then 0\n",
    "GDSC_mut_cnv_df = Cellline_mutation_df + cell_model_passport_cnv_T_df\n",
    "GDSC_mut_cnv_arr = np.where(GDSC_mut_cnv_df.values>=1,1,0)\n",
    "GDSC_mut_cnv_df = pd.DataFrame(index=GDSC_mut_cnv_df.index,columns=GDSC_mut_cnv_df.columns,data=GDSC_mut_cnv_arr)\n",
    "GDSC_mut_cnv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8de28ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "GDSC_mut_cnv_df.index = [str(sample) for sample in GDSC_mut_cnv_df.index]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a36a333",
   "metadata": {},
   "source": [
    "### 1.5 gene expression data from GDSC website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b5fc6788",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNAseq_GDSC_df = pd.read_csv(\"GDSC_original_data/EXPmerged.csv\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "e6b770e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000278619</th>\n",
       "      <th>ENSG00000278662</th>\n",
       "      <th>ENSG00000278704</th>\n",
       "      <th>ENSG00000278828</th>\n",
       "      <th>ENSG00000278845</th>\n",
       "      <th>ENSG00000280071</th>\n",
       "      <th>ENSG00000280433</th>\n",
       "      <th>ENSG00000280670</th>\n",
       "      <th>ENSG00000280789</th>\n",
       "      <th>ENSG00000280987</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1287381</th>\n",
       "      <td>5.742825</td>\n",
       "      <td>5.651001</td>\n",
       "      <td>4.467739</td>\n",
       "      <td>3.014166</td>\n",
       "      <td>-8.287284</td>\n",
       "      <td>7.213081</td>\n",
       "      <td>7.875378</td>\n",
       "      <td>7.328253</td>\n",
       "      <td>5.935010</td>\n",
       "      <td>3.913528</td>\n",
       "      <td>...</td>\n",
       "      <td>3.081384</td>\n",
       "      <td>1.984357</td>\n",
       "      <td>1.901941</td>\n",
       "      <td>-1.475922</td>\n",
       "      <td>5.893934</td>\n",
       "      <td>-0.778521</td>\n",
       "      <td>1.254001</td>\n",
       "      <td>0.463524</td>\n",
       "      <td>1.014700</td>\n",
       "      <td>3.523087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1287706</th>\n",
       "      <td>4.592277</td>\n",
       "      <td>5.974883</td>\n",
       "      <td>3.592532</td>\n",
       "      <td>5.099868</td>\n",
       "      <td>-5.318672</td>\n",
       "      <td>-3.159279</td>\n",
       "      <td>7.108233</td>\n",
       "      <td>4.926461</td>\n",
       "      <td>5.697733</td>\n",
       "      <td>3.430728</td>\n",
       "      <td>...</td>\n",
       "      <td>2.722252</td>\n",
       "      <td>0.274789</td>\n",
       "      <td>0.639703</td>\n",
       "      <td>-1.940189</td>\n",
       "      <td>4.372581</td>\n",
       "      <td>3.086884</td>\n",
       "      <td>0.500282</td>\n",
       "      <td>1.371473</td>\n",
       "      <td>2.619787</td>\n",
       "      <td>3.122676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910697</th>\n",
       "      <td>4.146911</td>\n",
       "      <td>6.570517</td>\n",
       "      <td>3.388971</td>\n",
       "      <td>2.786945</td>\n",
       "      <td>-1.732076</td>\n",
       "      <td>3.140121</td>\n",
       "      <td>6.143325</td>\n",
       "      <td>6.087315</td>\n",
       "      <td>6.018819</td>\n",
       "      <td>3.156478</td>\n",
       "      <td>...</td>\n",
       "      <td>2.799382</td>\n",
       "      <td>0.605541</td>\n",
       "      <td>-7.386796</td>\n",
       "      <td>-0.473561</td>\n",
       "      <td>4.977686</td>\n",
       "      <td>1.595625</td>\n",
       "      <td>0.968771</td>\n",
       "      <td>3.773703</td>\n",
       "      <td>3.666100</td>\n",
       "      <td>2.498545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910851</th>\n",
       "      <td>5.289480</td>\n",
       "      <td>5.993623</td>\n",
       "      <td>4.007116</td>\n",
       "      <td>4.056391</td>\n",
       "      <td>-3.789042</td>\n",
       "      <td>-2.565345</td>\n",
       "      <td>6.414339</td>\n",
       "      <td>4.548508</td>\n",
       "      <td>6.100855</td>\n",
       "      <td>3.962374</td>\n",
       "      <td>...</td>\n",
       "      <td>3.210041</td>\n",
       "      <td>1.059391</td>\n",
       "      <td>4.010217</td>\n",
       "      <td>-0.361289</td>\n",
       "      <td>6.496323</td>\n",
       "      <td>1.202010</td>\n",
       "      <td>1.153146</td>\n",
       "      <td>2.890040</td>\n",
       "      <td>2.867560</td>\n",
       "      <td>2.953157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910925</th>\n",
       "      <td>5.592378</td>\n",
       "      <td>6.567878</td>\n",
       "      <td>3.791862</td>\n",
       "      <td>5.487578</td>\n",
       "      <td>-2.851964</td>\n",
       "      <td>3.487037</td>\n",
       "      <td>7.073940</td>\n",
       "      <td>7.707461</td>\n",
       "      <td>5.538984</td>\n",
       "      <td>4.184304</td>\n",
       "      <td>...</td>\n",
       "      <td>3.447093</td>\n",
       "      <td>1.449922</td>\n",
       "      <td>0.460901</td>\n",
       "      <td>0.771684</td>\n",
       "      <td>5.855280</td>\n",
       "      <td>2.495384</td>\n",
       "      <td>0.577767</td>\n",
       "      <td>1.193354</td>\n",
       "      <td>3.131294</td>\n",
       "      <td>2.125035</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15370 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ENSG00000000003  ENSG00000000419  ENSG00000000457  ENSG00000000460  \\\n",
       "1287381         5.742825         5.651001         4.467739         3.014166   \n",
       "1287706         4.592277         5.974883         3.592532         5.099868   \n",
       "910697          4.146911         6.570517         3.388971         2.786945   \n",
       "910851          5.289480         5.993623         4.007116         4.056391   \n",
       "910925          5.592378         6.567878         3.791862         5.487578   \n",
       "\n",
       "         ENSG00000000938  ENSG00000000971  ENSG00000001036  ENSG00000001084  \\\n",
       "1287381        -8.287284         7.213081         7.875378         7.328253   \n",
       "1287706        -5.318672        -3.159279         7.108233         4.926461   \n",
       "910697         -1.732076         3.140121         6.143325         6.087315   \n",
       "910851         -3.789042        -2.565345         6.414339         4.548508   \n",
       "910925         -2.851964         3.487037         7.073940         7.707461   \n",
       "\n",
       "         ENSG00000001167  ENSG00000001460  ...  ENSG00000278619  \\\n",
       "1287381         5.935010         3.913528  ...         3.081384   \n",
       "1287706         5.697733         3.430728  ...         2.722252   \n",
       "910697          6.018819         3.156478  ...         2.799382   \n",
       "910851          6.100855         3.962374  ...         3.210041   \n",
       "910925          5.538984         4.184304  ...         3.447093   \n",
       "\n",
       "         ENSG00000278662  ENSG00000278704  ENSG00000278828  ENSG00000278845  \\\n",
       "1287381         1.984357         1.901941        -1.475922         5.893934   \n",
       "1287706         0.274789         0.639703        -1.940189         4.372581   \n",
       "910697          0.605541        -7.386796        -0.473561         4.977686   \n",
       "910851          1.059391         4.010217        -0.361289         6.496323   \n",
       "910925          1.449922         0.460901         0.771684         5.855280   \n",
       "\n",
       "         ENSG00000280071  ENSG00000280433  ENSG00000280670  ENSG00000280789  \\\n",
       "1287381        -0.778521         1.254001         0.463524         1.014700   \n",
       "1287706         3.086884         0.500282         1.371473         2.619787   \n",
       "910697          1.595625         0.968771         3.773703         3.666100   \n",
       "910851          1.202010         1.153146         2.890040         2.867560   \n",
       "910925          2.495384         0.577767         1.193354         3.131294   \n",
       "\n",
       "         ENSG00000280987  \n",
       "1287381         3.523087  \n",
       "1287706         3.122676  \n",
       "910697          2.498545  \n",
       "910851          2.953157  \n",
       "910925          2.125035  \n",
       "\n",
       "[5 rows x 15370 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_column_name = [sample.replace(\"cosmic.\",\"\") for sample in RNAseq_GDSC_df.columns]\n",
    "RNAseq_GDSC_df.columns = new_column_name\n",
    "RNAseq_GDSC_T_df = RNAseq_GDSC_df.T\n",
    "RNAseq_GDSC_T_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "d242f12b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(976, 15692)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000278619</th>\n",
       "      <th>ENSG00000278662</th>\n",
       "      <th>ENSG00000278704</th>\n",
       "      <th>ENSG00000278828</th>\n",
       "      <th>ENSG00000278845</th>\n",
       "      <th>ENSG00000280071</th>\n",
       "      <th>ENSG00000280433</th>\n",
       "      <th>ENSG00000280670</th>\n",
       "      <th>ENSG00000280789</th>\n",
       "      <th>ENSG00000280987</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>908132</th>\n",
       "      <td>-2.864894</td>\n",
       "      <td>5.085821</td>\n",
       "      <td>4.109730</td>\n",
       "      <td>4.658060</td>\n",
       "      <td>-4.069375</td>\n",
       "      <td>-4.387285</td>\n",
       "      <td>-0.690449</td>\n",
       "      <td>5.680146</td>\n",
       "      <td>6.059928</td>\n",
       "      <td>3.010456</td>\n",
       "      <td>...</td>\n",
       "      <td>4.459784</td>\n",
       "      <td>-0.297243</td>\n",
       "      <td>-1.126221</td>\n",
       "      <td>0.658029</td>\n",
       "      <td>5.208837</td>\n",
       "      <td>3.999111</td>\n",
       "      <td>2.632519</td>\n",
       "      <td>1.822507</td>\n",
       "      <td>2.894252</td>\n",
       "      <td>3.449887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924106</th>\n",
       "      <td>5.901626</td>\n",
       "      <td>5.900997</td>\n",
       "      <td>4.851388</td>\n",
       "      <td>5.570010</td>\n",
       "      <td>-5.287576</td>\n",
       "      <td>-5.109632</td>\n",
       "      <td>4.231945</td>\n",
       "      <td>6.141810</td>\n",
       "      <td>6.330790</td>\n",
       "      <td>3.601342</td>\n",
       "      <td>...</td>\n",
       "      <td>2.655478</td>\n",
       "      <td>-0.433687</td>\n",
       "      <td>-0.342844</td>\n",
       "      <td>0.211684</td>\n",
       "      <td>5.983458</td>\n",
       "      <td>1.518033</td>\n",
       "      <td>1.317567</td>\n",
       "      <td>1.143826</td>\n",
       "      <td>4.431685</td>\n",
       "      <td>1.186881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909729</th>\n",
       "      <td>3.756783</td>\n",
       "      <td>6.518574</td>\n",
       "      <td>2.671258</td>\n",
       "      <td>4.709470</td>\n",
       "      <td>-7.838378</td>\n",
       "      <td>-6.145803</td>\n",
       "      <td>7.815655</td>\n",
       "      <td>6.299586</td>\n",
       "      <td>5.890664</td>\n",
       "      <td>3.315878</td>\n",
       "      <td>...</td>\n",
       "      <td>1.994625</td>\n",
       "      <td>-0.761175</td>\n",
       "      <td>2.085720</td>\n",
       "      <td>0.101980</td>\n",
       "      <td>4.730947</td>\n",
       "      <td>-1.524343</td>\n",
       "      <td>0.801560</td>\n",
       "      <td>1.154081</td>\n",
       "      <td>1.880756</td>\n",
       "      <td>2.556203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907276</th>\n",
       "      <td>5.643208</td>\n",
       "      <td>6.569087</td>\n",
       "      <td>3.257356</td>\n",
       "      <td>4.484493</td>\n",
       "      <td>-5.152704</td>\n",
       "      <td>-4.656031</td>\n",
       "      <td>7.782905</td>\n",
       "      <td>7.938462</td>\n",
       "      <td>5.241121</td>\n",
       "      <td>3.460099</td>\n",
       "      <td>...</td>\n",
       "      <td>4.619435</td>\n",
       "      <td>0.625857</td>\n",
       "      <td>2.331522</td>\n",
       "      <td>1.196622</td>\n",
       "      <td>5.845032</td>\n",
       "      <td>0.971824</td>\n",
       "      <td>0.031280</td>\n",
       "      <td>1.010198</td>\n",
       "      <td>3.586525</td>\n",
       "      <td>1.486064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946363</th>\n",
       "      <td>5.243715</td>\n",
       "      <td>5.622904</td>\n",
       "      <td>3.815669</td>\n",
       "      <td>5.111991</td>\n",
       "      <td>-6.986853</td>\n",
       "      <td>-1.020738</td>\n",
       "      <td>3.459126</td>\n",
       "      <td>5.805482</td>\n",
       "      <td>6.247374</td>\n",
       "      <td>2.247746</td>\n",
       "      <td>...</td>\n",
       "      <td>1.900902</td>\n",
       "      <td>1.794033</td>\n",
       "      <td>1.144987</td>\n",
       "      <td>-0.866314</td>\n",
       "      <td>5.482496</td>\n",
       "      <td>-0.184195</td>\n",
       "      <td>-0.116829</td>\n",
       "      <td>2.725201</td>\n",
       "      <td>2.984674</td>\n",
       "      <td>2.860856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15370 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ENSG00000000003  ENSG00000000419  ENSG00000000457  ENSG00000000460  \\\n",
       "908132        -2.864894         5.085821         4.109730         4.658060   \n",
       "924106         5.901626         5.900997         4.851388         5.570010   \n",
       "909729         3.756783         6.518574         2.671258         4.709470   \n",
       "907276         5.643208         6.569087         3.257356         4.484493   \n",
       "946363         5.243715         5.622904         3.815669         5.111991   \n",
       "\n",
       "        ENSG00000000938  ENSG00000000971  ENSG00000001036  ENSG00000001084  \\\n",
       "908132        -4.069375        -4.387285        -0.690449         5.680146   \n",
       "924106        -5.287576        -5.109632         4.231945         6.141810   \n",
       "909729        -7.838378        -6.145803         7.815655         6.299586   \n",
       "907276        -5.152704        -4.656031         7.782905         7.938462   \n",
       "946363        -6.986853        -1.020738         3.459126         5.805482   \n",
       "\n",
       "        ENSG00000001167  ENSG00000001460  ...  ENSG00000278619  \\\n",
       "908132         6.059928         3.010456  ...         4.459784   \n",
       "924106         6.330790         3.601342  ...         2.655478   \n",
       "909729         5.890664         3.315878  ...         1.994625   \n",
       "907276         5.241121         3.460099  ...         4.619435   \n",
       "946363         6.247374         2.247746  ...         1.900902   \n",
       "\n",
       "        ENSG00000278662  ENSG00000278704  ENSG00000278828  ENSG00000278845  \\\n",
       "908132        -0.297243        -1.126221         0.658029         5.208837   \n",
       "924106        -0.433687        -0.342844         0.211684         5.983458   \n",
       "909729        -0.761175         2.085720         0.101980         4.730947   \n",
       "907276         0.625857         2.331522         1.196622         5.845032   \n",
       "946363         1.794033         1.144987        -0.866314         5.482496   \n",
       "\n",
       "        ENSG00000280071  ENSG00000280433  ENSG00000280670  ENSG00000280789  \\\n",
       "908132         3.999111         2.632519         1.822507         2.894252   \n",
       "924106         1.518033         1.317567         1.143826         4.431685   \n",
       "909729        -1.524343         0.801560         1.154081         1.880756   \n",
       "907276         0.971824         0.031280         1.010198         3.586525   \n",
       "946363        -0.184195        -0.116829         2.725201         2.984674   \n",
       "\n",
       "        ENSG00000280987  \n",
       "908132         3.449887  \n",
       "924106         1.186881  \n",
       "909729         2.556203  \n",
       "907276         1.486064  \n",
       "946363         2.860856  \n",
       "\n",
       "[5 rows x 15370 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GDSC_common_sample = list(set(GDSC_mut_cnv_df.index)&set(RNAseq_GDSC_T_df.index))\n",
    "RNAseq_GDSC_T_df = RNAseq_GDSC_T_df.loc[GDSC_common_sample ]\n",
    "GDSC_mut_cnv_df = GDSC_mut_cnv_df.loc[GDSC_common_sample ]\n",
    "print(GDSC_mut_cnv_df.shape)\n",
    "RNAseq_GDSC_T_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "a26a307d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:biothings.client:querying 1-1000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 1001-2000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 2001-3000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 3001-4000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 4001-5000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 5001-6000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 6001-7000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 7001-8000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 8001-9000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 9001-10000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 10001-11000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 11001-12000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 12001-13000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 13001-14000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 14001-15000...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:querying 15001-15370...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:Finished.\n",
      "WARNING:biothings.client:24 input query terms found no hit:\t['ENSG00000069712', 'ENSG00000112096', 'ENSG00000116957', 'ENSG00000130489', 'ENSG00000130723', 'ENS\n",
      "INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"
     ]
    }
   ],
   "source": [
    "#change the gene name from ensemble name to gene symbol, different version has some variance to define the gene name\n",
    "mg = mygene.MyGeneInfo()\n",
    "geneList = RNAseq_GDSC_T_df.columns\n",
    "geneSyms = mg.querymany(geneList , scopes='ensembl.gene', fields='symbol', species='human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "126ceaa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>FIRRM</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>STPG1</th>\n",
       "      <th>...</th>\n",
       "      <th>C11orf98</th>\n",
       "      <th>MRM1</th>\n",
       "      <th>GOLGA6L10</th>\n",
       "      <th>H3C10</th>\n",
       "      <th>MRPL45</th>\n",
       "      <th>LOC102724023</th>\n",
       "      <th>LOC102724200</th>\n",
       "      <th>CCDC163</th>\n",
       "      <th>PAGR1</th>\n",
       "      <th>MATR3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>908132</th>\n",
       "      <td>-2.864894</td>\n",
       "      <td>5.085821</td>\n",
       "      <td>4.109730</td>\n",
       "      <td>4.658060</td>\n",
       "      <td>-4.069375</td>\n",
       "      <td>-4.387285</td>\n",
       "      <td>-0.690449</td>\n",
       "      <td>5.680146</td>\n",
       "      <td>6.059928</td>\n",
       "      <td>3.010456</td>\n",
       "      <td>...</td>\n",
       "      <td>2.791495</td>\n",
       "      <td>4.459784</td>\n",
       "      <td>-0.297243</td>\n",
       "      <td>0.658029</td>\n",
       "      <td>5.208837</td>\n",
       "      <td>3.999111</td>\n",
       "      <td>2.632519</td>\n",
       "      <td>1.822507</td>\n",
       "      <td>2.894252</td>\n",
       "      <td>3.449887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924106</th>\n",
       "      <td>5.901626</td>\n",
       "      <td>5.900997</td>\n",
       "      <td>4.851388</td>\n",
       "      <td>5.570010</td>\n",
       "      <td>-5.287576</td>\n",
       "      <td>-5.109632</td>\n",
       "      <td>4.231945</td>\n",
       "      <td>6.141810</td>\n",
       "      <td>6.330790</td>\n",
       "      <td>3.601342</td>\n",
       "      <td>...</td>\n",
       "      <td>2.759417</td>\n",
       "      <td>2.655478</td>\n",
       "      <td>-0.433687</td>\n",
       "      <td>0.211684</td>\n",
       "      <td>5.983458</td>\n",
       "      <td>1.518033</td>\n",
       "      <td>1.317567</td>\n",
       "      <td>1.143826</td>\n",
       "      <td>4.431685</td>\n",
       "      <td>1.186881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909729</th>\n",
       "      <td>3.756783</td>\n",
       "      <td>6.518574</td>\n",
       "      <td>2.671258</td>\n",
       "      <td>4.709470</td>\n",
       "      <td>-7.838378</td>\n",
       "      <td>-6.145803</td>\n",
       "      <td>7.815655</td>\n",
       "      <td>6.299586</td>\n",
       "      <td>5.890664</td>\n",
       "      <td>3.315878</td>\n",
       "      <td>...</td>\n",
       "      <td>3.775111</td>\n",
       "      <td>1.994625</td>\n",
       "      <td>-0.761175</td>\n",
       "      <td>0.101980</td>\n",
       "      <td>4.730947</td>\n",
       "      <td>-1.524343</td>\n",
       "      <td>0.801560</td>\n",
       "      <td>1.154081</td>\n",
       "      <td>1.880756</td>\n",
       "      <td>2.556203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907276</th>\n",
       "      <td>5.643208</td>\n",
       "      <td>6.569087</td>\n",
       "      <td>3.257356</td>\n",
       "      <td>4.484493</td>\n",
       "      <td>-5.152704</td>\n",
       "      <td>-4.656031</td>\n",
       "      <td>7.782905</td>\n",
       "      <td>7.938462</td>\n",
       "      <td>5.241121</td>\n",
       "      <td>3.460099</td>\n",
       "      <td>...</td>\n",
       "      <td>3.753014</td>\n",
       "      <td>4.619435</td>\n",
       "      <td>0.625857</td>\n",
       "      <td>1.196622</td>\n",
       "      <td>5.845032</td>\n",
       "      <td>0.971824</td>\n",
       "      <td>0.031280</td>\n",
       "      <td>1.010198</td>\n",
       "      <td>3.586525</td>\n",
       "      <td>1.486064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946363</th>\n",
       "      <td>5.243715</td>\n",
       "      <td>5.622904</td>\n",
       "      <td>3.815669</td>\n",
       "      <td>5.111991</td>\n",
       "      <td>-6.986853</td>\n",
       "      <td>-1.020738</td>\n",
       "      <td>3.459126</td>\n",
       "      <td>5.805482</td>\n",
       "      <td>6.247374</td>\n",
       "      <td>2.247746</td>\n",
       "      <td>...</td>\n",
       "      <td>2.730621</td>\n",
       "      <td>1.900902</td>\n",
       "      <td>1.794033</td>\n",
       "      <td>-0.866314</td>\n",
       "      <td>5.482496</td>\n",
       "      <td>-0.184195</td>\n",
       "      <td>-0.116829</td>\n",
       "      <td>2.725201</td>\n",
       "      <td>2.984674</td>\n",
       "      <td>2.860856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15323 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TSPAN6      DPM1     SCYL3     FIRRM       FGR       CFH     FUCA2  \\\n",
       "908132 -2.864894  5.085821  4.109730  4.658060 -4.069375 -4.387285 -0.690449   \n",
       "924106  5.901626  5.900997  4.851388  5.570010 -5.287576 -5.109632  4.231945   \n",
       "909729  3.756783  6.518574  2.671258  4.709470 -7.838378 -6.145803  7.815655   \n",
       "907276  5.643208  6.569087  3.257356  4.484493 -5.152704 -4.656031  7.782905   \n",
       "946363  5.243715  5.622904  3.815669  5.111991 -6.986853 -1.020738  3.459126   \n",
       "\n",
       "            GCLC      NFYA     STPG1  ...  C11orf98      MRM1  GOLGA6L10  \\\n",
       "908132  5.680146  6.059928  3.010456  ...  2.791495  4.459784  -0.297243   \n",
       "924106  6.141810  6.330790  3.601342  ...  2.759417  2.655478  -0.433687   \n",
       "909729  6.299586  5.890664  3.315878  ...  3.775111  1.994625  -0.761175   \n",
       "907276  7.938462  5.241121  3.460099  ...  3.753014  4.619435   0.625857   \n",
       "946363  5.805482  6.247374  2.247746  ...  2.730621  1.900902   1.794033   \n",
       "\n",
       "           H3C10    MRPL45  LOC102724023  LOC102724200   CCDC163     PAGR1  \\\n",
       "908132  0.658029  5.208837      3.999111      2.632519  1.822507  2.894252   \n",
       "924106  0.211684  5.983458      1.518033      1.317567  1.143826  4.431685   \n",
       "909729  0.101980  4.730947     -1.524343      0.801560  1.154081  1.880756   \n",
       "907276  1.196622  5.845032      0.971824      0.031280  1.010198  3.586525   \n",
       "946363 -0.866314  5.482496     -0.184195     -0.116829  2.725201  2.984674   \n",
       "\n",
       "           MATR3  \n",
       "908132  3.449887  \n",
       "924106  1.186881  \n",
       "909729  2.556203  \n",
       "907276  1.486064  \n",
       "946363  2.860856  \n",
       "\n",
       "[5 rows x 15323 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_gene_name = []\n",
    "gene_symbol_name = []\n",
    "for item in geneSyms:\n",
    "    if 'symbol' in item:\n",
    "        ensembl_gene_name.append(item['query'])\n",
    "        gene_symbol_name.append(item['symbol'])\n",
    "RNAseq_GDSC_T_df = RNAseq_GDSC_T_df[ensembl_gene_name]\n",
    "RNAseq_GDSC_T_df.columns = gene_symbol_name\n",
    "RNAseq_GDSC_T_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f7c21a3",
   "metadata": {},
   "source": [
    "### 1.6 cancer type data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a7ca9d8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "908132     NaN\n",
       "924106    BRCA\n",
       "909729     GBM\n",
       "907276    STAD\n",
       "946363      NB\n",
       "          ... \n",
       "688021    SCLC\n",
       "971774     NaN\n",
       "909255     NaN\n",
       "949170      NB\n",
       "906763     ALL\n",
       "Name: cancer_type, Length: 976, dtype: object"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GDSC_cancer_type_df = Cellline_mutation_cancer_type_df.iloc[:,-1]\n",
    "GDSC_cancer_type_df.index = [str(sample) for sample in GDSC_cancer_type_df.index]\n",
    "GDSC_cancer_type_df = GDSC_cancer_type_df.loc[RNAseq_GDSC_T_df.index]\n",
    "GDSC_cancer_type_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0bad9ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6163e015",
   "metadata": {},
   "source": [
    "# 2. TCGA dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "769f932e",
   "metadata": {},
   "source": [
    "### 2.1 mutation data from TCGA website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e31ddf16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9104, 40543)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>UBE2Q2</th>\n",
       "      <th>CHMP1B</th>\n",
       "      <th>PSMA2P1</th>\n",
       "      <th>SHQ1P1</th>\n",
       "      <th>CPHL1P</th>\n",
       "      <th>SSXP10</th>\n",
       "      <th>REM1</th>\n",
       "      <th>TCOF1</th>\n",
       "      <th>NSRP1</th>\n",
       "      <th>OPA6</th>\n",
       "      <th>...</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>OR1E5</th>\n",
       "      <th>RP11-390F4.3</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>PTRF</th>\n",
       "      <th>DIAPH2-AS1</th>\n",
       "      <th>SELV</th>\n",
       "      <th>NFIX</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0003-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0033-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2470-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40543 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample           UBE2Q2  CHMP1B  PSMA2P1  SHQ1P1  CPHL1P  SSXP10  REM1  TCOF1  \\\n",
       "TCGA-02-0003-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-0033-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-0047-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-0055-01       0       0        0       0       0       0     0      0   \n",
       "TCGA-02-2470-01       0       0        0       0       0       0     0      0   \n",
       "\n",
       "sample           NSRP1  OPA6  ...  TULP2  OR1E5  RP11-390F4.3  GNGT2  GNGT1  \\\n",
       "TCGA-02-0003-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-0033-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-0047-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-0055-01      0     0  ...      0      0             0      0      0   \n",
       "TCGA-02-2470-01      0     0  ...      0      0             0      0      0   \n",
       "\n",
       "sample           PTRF  DIAPH2-AS1  SELV  NFIX  SELP  \n",
       "TCGA-02-0003-01     0           0     0     0     0  \n",
       "TCGA-02-0033-01     0           0     0     0     0  \n",
       "TCGA-02-0047-01     0           0     0     0     0  \n",
       "TCGA-02-0055-01     0           0     0     0     0  \n",
       "TCGA-02-2470-01     0           0     0     1     0  \n",
       "\n",
       "[5 rows x 40543 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_mutation_df = pd.read_csv(\"TCGA_original_data/mc3.v0.2.8.PUBLIC.nonsilentGene.xena\",sep = \"\\t\",header = 0, index_col = 0, keep_default_na = False) # Downloaded from the UCSC Xena TCGA data portal\n",
    "TCGA_mutation_df = TCGA_mutation_df.T\n",
    "print(TCGA_mutation_df.shape)\n",
    "TCGA_mutation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0d165001",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9104, 40543)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_mutation_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ae2d669",
   "metadata": {},
   "source": [
    "### 2.2 CNV data from TCGA website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8110f5bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_cnv_ori_df = pd.read_csv(\"TCGA_original_data/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes\",sep='\\t',index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c2cb0e7",
   "metadata": {},
   "source": [
    "### 2.3 SGA data processing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ce6629b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_cnv_df = TCGA_cnv_ori_df.T\n",
    "TCGA_cnv_df = TCGA_cnv_df.replace([-2,-1,0,1,2],[1,0,0,0,1]) # for cnv data, only set -2 and 2 to be 1, otherwise 0\n",
    "TCGA_common_samples = list(set(TCGA_mutation_df.index)&set(TCGA_cnv_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3221a8d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>SDHB</th>\n",
       "      <th>PPP1R14D</th>\n",
       "      <th>ZNF66</th>\n",
       "      <th>ATXN7L3</th>\n",
       "      <th>MIR4263</th>\n",
       "      <th>SYTL1</th>\n",
       "      <th>RELL1</th>\n",
       "      <th>HECTD4</th>\n",
       "      <th>IL36G</th>\n",
       "      <th>LINC00621</th>\n",
       "      <th>...</th>\n",
       "      <th>UBE2Q2P2</th>\n",
       "      <th>LPHN3</th>\n",
       "      <th>RN7SL862P</th>\n",
       "      <th>RGS16</th>\n",
       "      <th>C4orf40</th>\n",
       "      <th>ADAT1</th>\n",
       "      <th>PEBP1</th>\n",
       "      <th>CDKL1</th>\n",
       "      <th>TRAM2</th>\n",
       "      <th>CHRNA6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-78-7542-01</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-G9-6367-01</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-S2-AA1A-01</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9JG-01</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LO-01</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23384 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample           SDHB  PPP1R14D  ZNF66  ATXN7L3  MIR4263  SYTL1  RELL1  \\\n",
       "TCGA-78-7542-01   0.0       0.0    0.0      0.0      0.0    0.0    0.0   \n",
       "TCGA-G9-6367-01   0.0       0.0    0.0      0.0      0.0    0.0    0.0   \n",
       "TCGA-S2-AA1A-01   0.0       0.0    0.0      0.0      0.0    0.0    0.0   \n",
       "TCGA-2Z-A9JG-01   0.0       0.0    0.0      0.0      0.0    0.0    0.0   \n",
       "TCGA-OR-A5LO-01   0.0       0.0    0.0      1.0      0.0    0.0    0.0   \n",
       "\n",
       "Sample           HECTD4  IL36G  LINC00621  ...  UBE2Q2P2  LPHN3  RN7SL862P  \\\n",
       "TCGA-78-7542-01     0.0    0.0        0.0  ...       0.0    0.0        0.0   \n",
       "TCGA-G9-6367-01     0.0    0.0        0.0  ...       0.0    0.0        0.0   \n",
       "TCGA-S2-AA1A-01     0.0    0.0        0.0  ...       0.0    0.0        0.0   \n",
       "TCGA-2Z-A9JG-01     0.0    0.0        0.0  ...       0.0    0.0        0.0   \n",
       "TCGA-OR-A5LO-01     0.0    0.0        0.0  ...       0.0    0.0        0.0   \n",
       "\n",
       "Sample           RGS16  C4orf40  ADAT1  PEBP1  CDKL1  TRAM2  CHRNA6  \n",
       "TCGA-78-7542-01    0.0      0.0    0.0    0.0    0.0    0.0     0.0  \n",
       "TCGA-G9-6367-01    0.0      0.0    0.0    0.0    0.0    0.0     0.0  \n",
       "TCGA-S2-AA1A-01    0.0      0.0    0.0    0.0    0.0    0.0     0.0  \n",
       "TCGA-2Z-A9JG-01    0.0      0.0    0.0    0.0    0.0    0.0     0.0  \n",
       "TCGA-OR-A5LO-01    0.0      0.0    0.0    0.0    0.0    0.0     0.0  \n",
       "\n",
       "[5 rows x 23384 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_common_gene_mut_cnv_list = list(set(TCGA_mutation_df.columns)&set(TCGA_cnv_df.columns))\n",
    "TCGA_cnv_df = TCGA_cnv_df.loc[TCGA_common_samples,TCGA_common_gene_mut_cnv_list]\n",
    "TCGA_mutation_df = TCGA_mutation_df.loc[TCGA_common_samples,TCGA_common_gene_mut_cnv_list]\n",
    "\n",
    "TCGA_mutation_df = TCGA_mutation_df.astype(float)\n",
    "TCGA_mut_cnv_df = TCGA_cnv_df + TCGA_mutation_df\n",
    "TCGA_mut_cnv_df = TCGA_mut_cnv_df.replace([0,1,2],[0,1,1])\n",
    "TCGA_mut_cnv_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "add41c32",
   "metadata": {},
   "source": [
    "### 2.4 expression data from TCGA website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a81647a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-OR-A5J1-01</th>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <th>TCGA-OR-A5J3-01</th>\n",
       "      <th>TCGA-OR-A5J5-01</th>\n",
       "      <th>TCGA-OR-A5J6-01</th>\n",
       "      <th>TCGA-OR-A5J7-01</th>\n",
       "      <th>TCGA-OR-A5J8-01</th>\n",
       "      <th>TCGA-OR-A5J9-01</th>\n",
       "      <th>TCGA-OR-A5JA-01</th>\n",
       "      <th>TCGA-OR-A5JB-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-CG-4449-01</th>\n",
       "      <th>TCGA-CG-4462-01</th>\n",
       "      <th>TCGA-CG-4465-01</th>\n",
       "      <th>TCGA-CG-4466-01</th>\n",
       "      <th>TCGA-CG-4469-01</th>\n",
       "      <th>TCGA-CG-4472-01</th>\n",
       "      <th>TCGA-CG-4474-01</th>\n",
       "      <th>TCGA-CG-4475-01</th>\n",
       "      <th>TCGA-CG-4476-01</th>\n",
       "      <th>TCGA-CG-4477-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100130426</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100133144</th>\n",
       "      <td>2.09</td>\n",
       "      <td>1.88</td>\n",
       "      <td>1.45</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.12</td>\n",
       "      <td>1.29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.45</td>\n",
       "      <td>3.02</td>\n",
       "      <td>...</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.74</td>\n",
       "      <td>2.64</td>\n",
       "      <td>3.99</td>\n",
       "      <td>4.40</td>\n",
       "      <td>3.00</td>\n",
       "      <td>4.27</td>\n",
       "      <td>3.71</td>\n",
       "      <td>4.29</td>\n",
       "      <td>4.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100134869</th>\n",
       "      <td>2.30</td>\n",
       "      <td>3.32</td>\n",
       "      <td>2.92</td>\n",
       "      <td>1.35</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2.82</td>\n",
       "      <td>1.72</td>\n",
       "      <td>1.24</td>\n",
       "      <td>2.60</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.87</td>\n",
       "      <td>2.12</td>\n",
       "      <td>1.78</td>\n",
       "      <td>2.60</td>\n",
       "      <td>3.66</td>\n",
       "      <td>3.87</td>\n",
       "      <td>3.07</td>\n",
       "      <td>3.65</td>\n",
       "      <td>3.64</td>\n",
       "      <td>3.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10357</th>\n",
       "      <td>7.23</td>\n",
       "      <td>6.36</td>\n",
       "      <td>6.45</td>\n",
       "      <td>5.78</td>\n",
       "      <td>6.09</td>\n",
       "      <td>6.71</td>\n",
       "      <td>6.58</td>\n",
       "      <td>6.31</td>\n",
       "      <td>6.13</td>\n",
       "      <td>7.29</td>\n",
       "      <td>...</td>\n",
       "      <td>9.31</td>\n",
       "      <td>8.20</td>\n",
       "      <td>9.43</td>\n",
       "      <td>9.14</td>\n",
       "      <td>10.02</td>\n",
       "      <td>9.32</td>\n",
       "      <td>9.54</td>\n",
       "      <td>8.99</td>\n",
       "      <td>9.48</td>\n",
       "      <td>9.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10431</th>\n",
       "      <td>10.99</td>\n",
       "      <td>10.35</td>\n",
       "      <td>10.04</td>\n",
       "      <td>11.20</td>\n",
       "      <td>10.30</td>\n",
       "      <td>10.87</td>\n",
       "      <td>9.96</td>\n",
       "      <td>10.78</td>\n",
       "      <td>10.25</td>\n",
       "      <td>10.97</td>\n",
       "      <td>...</td>\n",
       "      <td>10.23</td>\n",
       "      <td>9.33</td>\n",
       "      <td>10.21</td>\n",
       "      <td>9.38</td>\n",
       "      <td>9.34</td>\n",
       "      <td>10.34</td>\n",
       "      <td>10.17</td>\n",
       "      <td>10.43</td>\n",
       "      <td>9.93</td>\n",
       "      <td>10.76</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 11069 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TCGA-OR-A5J1-01  TCGA-OR-A5J2-01  TCGA-OR-A5J3-01  TCGA-OR-A5J5-01  \\\n",
       "sample                                                                          \n",
       "100130426             0.00             0.00             0.00             0.00   \n",
       "100133144             2.09             1.88             1.45             0.00   \n",
       "100134869             2.30             3.32             2.92             1.35   \n",
       "10357                 7.23             6.36             6.45             5.78   \n",
       "10431                10.99            10.35            10.04            11.20   \n",
       "\n",
       "           TCGA-OR-A5J6-01  TCGA-OR-A5J7-01  TCGA-OR-A5J8-01  TCGA-OR-A5J9-01  \\\n",
       "sample                                                                          \n",
       "100130426             0.00             0.00             0.00             0.00   \n",
       "100133144             0.00             1.12             1.29             0.00   \n",
       "100134869             2.45             2.82             1.72             1.24   \n",
       "10357                 6.09             6.71             6.58             6.31   \n",
       "10431                10.30            10.87             9.96            10.78   \n",
       "\n",
       "           TCGA-OR-A5JA-01  TCGA-OR-A5JB-01  ...  TCGA-CG-4449-01  \\\n",
       "sample                                       ...                    \n",
       "100130426             0.00             0.00  ...              NaN   \n",
       "100133144             2.45             3.02  ...             2.42   \n",
       "100134869             2.60             0.00  ...             1.87   \n",
       "10357                 6.13             7.29  ...             9.31   \n",
       "10431                10.25            10.97  ...            10.23   \n",
       "\n",
       "           TCGA-CG-4462-01  TCGA-CG-4465-01  TCGA-CG-4466-01  TCGA-CG-4469-01  \\\n",
       "sample                                                                          \n",
       "100130426              NaN              NaN              NaN              NaN   \n",
       "100133144             2.74             2.64             3.99             4.40   \n",
       "100134869             2.12             1.78             2.60             3.66   \n",
       "10357                 8.20             9.43             9.14            10.02   \n",
       "10431                 9.33            10.21             9.38             9.34   \n",
       "\n",
       "           TCGA-CG-4472-01  TCGA-CG-4474-01  TCGA-CG-4475-01  TCGA-CG-4476-01  \\\n",
       "sample                                                                          \n",
       "100130426              NaN              NaN              NaN              NaN   \n",
       "100133144             3.00             4.27             3.71             4.29   \n",
       "100134869             3.87             3.07             3.65             3.64   \n",
       "10357                 9.32             9.54             8.99             9.48   \n",
       "10431                10.34            10.17            10.43             9.93   \n",
       "\n",
       "           TCGA-CG-4477-01  \n",
       "sample                      \n",
       "100130426              NaN  \n",
       "100133144             4.24  \n",
       "100134869             3.99  \n",
       "10357                 9.46  \n",
       "10431                10.76  \n",
       "\n",
       "[5 rows x 11069 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_xena_exprs_df = pd.read_csv(\"TCGA_original_data/EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena\",sep=\"\\t\",index_col=0) # Downloaded from the UCSC Xena TCGA data portal\n",
    "TCGA_xena_exprs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b0723f2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11069, 16335)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_exprs_df = TCGA_xena_exprs_df.dropna().T\n",
    "TCGA_exprs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4344597e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8893, 23384)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_mut_cnv_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e561fd86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11069, 16335)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_exprs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "85c1bd9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8586, 16335) (8586, 23384)\n"
     ]
    }
   ],
   "source": [
    "TCGA_common_sample_exprs_sga = list(set(TCGA_exprs_df.index)&set(TCGA_mut_cnv_df.index))\n",
    "TCGA_exprs_processed_df = TCGA_exprs_df.loc[TCGA_common_sample_exprs_sga]\n",
    "TCGA_mut_cnv_processed_df = TCGA_mut_cnv_df.loc[TCGA_common_sample_exprs_sga]\n",
    "print(TCGA_exprs_processed_df.shape,TCGA_mut_cnv_processed_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd54dc54",
   "metadata": {},
   "source": [
    "### 2.5 cancer type data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "edf548fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study Abbreviation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>acute myeloid leukemia</th>\n",
       "      <td>LAML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>adrenocortical cancer</th>\n",
       "      <td>ACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bladder urothelial carcinoma</th>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brain lower grade glioma</th>\n",
       "      <td>LGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>breast invasive carcinoma</th>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Study Abbreviation\n",
       "acute myeloid leukemia                     LAML\n",
       "adrenocortical cancer                       ACC\n",
       "bladder urothelial carcinoma               BLCA\n",
       "brain lower grade glioma                    LGG\n",
       "breast invasive carcinoma                  BRCA"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_Study_Abbreviations_df = pd.read_excel(\"TCGA_original_data/TCGA_Study_Abbreviations.xlsx\",index_col=1)\n",
    "TCGA_Study_Abbreviations_new_index = [sample.lower() for sample in TCGA_Study_Abbreviations_df.index]\n",
    "TCGA_Study_Abbreviations_new_index = [disease[:-1] if disease.endswith('s') else disease for disease in TCGA_Study_Abbreviations_new_index]\n",
    "for i, element in enumerate(TCGA_Study_Abbreviations_new_index):\n",
    "    if element == \"uterine corpus endometrial carcinoma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = \"uterine corpus endometrioid carcinoma\"\n",
    "    if element == \"cervical squamous cell carcinoma and endocervical adenocarcinoma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = 'cervical & endocervical cancer'\n",
    "    if element == \"pheochromocytoma and paraganglioma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = 'pheochromocytoma & paraganglioma'\n",
    "    if element == \"head and neck squamous cell carcinoma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = 'head & neck squamous cell carcinoma'\n",
    "    if element == \"kidney renal papillary cell carcinoma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = 'kidney papillary cell carcinoma'\n",
    "    if element == \"kidney renal clear cell carcinoma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = 'kidney clear cell carcinoma'\n",
    "    if element == \"lymphoid neoplasm diffuse large b-cell lymphoma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = 'diffuse large B-cell lymphoma'\n",
    "    if element == \"adrenocortical carcinoma\":\n",
    "        TCGA_Study_Abbreviations_new_index[i] = 'adrenocortical cancer'\n",
    "TCGA_Study_Abbreviations_df.index = TCGA_Study_Abbreviations_new_index \n",
    "\n",
    "TCGA_Study_Abbreviations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a699d6bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-78-7542-01</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-G9-6367-01</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-S2-AA1A-01</th>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9JG-01</th>\n",
       "      <td>KIRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5LO-01</th>\n",
       "      <td>ACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-XJ-A83H-01</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-E8-A2EA-01</th>\n",
       "      <td>THCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-XJ-A9DQ-01</th>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-FD-A3SP-01</th>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-TM-A84B-01</th>\n",
       "      <td>LGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8586 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                cancer_type\n",
       "sample                     \n",
       "TCGA-78-7542-01        LUAD\n",
       "TCGA-G9-6367-01        PRAD\n",
       "TCGA-S2-AA1A-01        LUAD\n",
       "TCGA-2Z-A9JG-01        KIRP\n",
       "TCGA-OR-A5LO-01         ACC\n",
       "...                     ...\n",
       "TCGA-XJ-A83H-01        PRAD\n",
       "TCGA-E8-A2EA-01        THCA\n",
       "TCGA-XJ-A9DQ-01        PRAD\n",
       "TCGA-FD-A3SP-01        BLCA\n",
       "TCGA-TM-A84B-01         LGG\n",
       "\n",
       "[8586 rows x 1 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_xena_cancertype = pd.read_csv(\"TCGA_original_data/TCGA_phenotype_denseDataOnlyDownload.tsv\",sep=\"\\t\",index_col=0)\n",
    "cancer_type = TCGA_Study_Abbreviations_df.loc[TCGA_xena_cancertype.loc[:,\"_primary_disease\"],\"Study Abbreviation\"].to_list()\n",
    "TCGA_xena_cancertype[\"cancer_type\"] = cancer_type\n",
    "TCGA_cancer_type = TCGA_xena_cancertype.loc[TCGA_common_sample_exprs_sga,\"cancer_type\"].to_frame()\n",
    "TCGA_cancer_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "705bfe86",
   "metadata": {},
   "source": [
    "# 3. Combine GDSC and TCGA dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a6fd849",
   "metadata": {},
   "source": [
    "### 3.1 expression data combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4dd353f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12519\n",
      "(976, 12519) (8586, 12519)\n"
     ]
    }
   ],
   "source": [
    "RNAseq_common_gene_gdsc_tcga = list(set(RNAseq_GDSC_T_df.columns)&set(TCGA_exprs_processed_df.columns))\n",
    "print(len(RNAseq_common_gene_gdsc_tcga))\n",
    "\n",
    "RNAseq_GDSC_T_df = RNAseq_GDSC_T_df[RNAseq_common_gene_gdsc_tcga]\n",
    "RNAseq_GDSC_final_df = RNAseq_GDSC_T_df.loc[:,~RNAseq_GDSC_T_df.columns.duplicated()].copy() #drop duplicate column\n",
    "TCGA_exprs_final_df = TCGA_exprs_processed_df[RNAseq_common_gene_gdsc_tcga]\n",
    "print(RNAseq_GDSC_final_df.shape,TCGA_exprs_final_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f0248306",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            0\n",
      "0        A1BG\n",
      "1        A1CF\n",
      "2         A2M\n",
      "3       AADAC\n",
      "4      ABCA12\n",
      "...       ...\n",
      "2753    ZNF91\n",
      "2754    ZNF92\n",
      "2755      ZP3\n",
      "2756  ZSCAN16\n",
      "2757  ZSCAN18\n",
      "\n",
      "[2758 rows x 1 columns]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2317"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_exprs_list = pd.read_csv(\"exprs_gene_list_GDSC_yifan_mike.txt\",header=None)\n",
    "common_gene_exprs = list(set(RNAseq_GDSC_final_df.columns)&set(gene_exprs_list.iloc[:,0].tolist()))\n",
    "len(common_gene_exprs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48f889cb",
   "metadata": {},
   "source": [
    "# Since the package \"mygene\" change the version with time, the gene name has a little change, this will result to different number of overlap genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "bf63aca3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(976, 2317) (8586, 2317)\n"
     ]
    }
   ],
   "source": [
    "GDSC_RNAseq_df = RNAseq_GDSC_final_df[common_gene_exprs]\n",
    "TCGA_RNAseq_df = TCGA_exprs_final_df[common_gene_exprs]\n",
    "print(GDSC_RNAseq_df.shape,TCGA_RNAseq_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a22595a3",
   "metadata": {},
   "source": [
    "### 3.2 SGA data combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c01c81e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(976, 15692) (8586, 15692)\n"
     ]
    }
   ],
   "source": [
    "SGA_common_gene_gdsc_tcga = list(set(TCGA_mut_cnv_processed_df.columns)&set(GDSC_mut_cnv_df.columns))\n",
    "GDSC_mut_cnv_df = GDSC_mut_cnv_df[SGA_common_gene_gdsc_tcga]\n",
    "TCGA_mut_cnv_processed_df = TCGA_mut_cnv_processed_df[SGA_common_gene_gdsc_tcga]\n",
    "print(GDSC_mut_cnv_df.shape,TCGA_mut_cnv_processed_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a8d44247",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1084\n"
     ]
    }
   ],
   "source": [
    "gene_sga_list = pd.read_csv(\"TCI_fondation1_driver_dataset_combine_gene_list.txt\",header=None)\n",
    "common_gene_sga = list(set(GDSC_mut_cnv_df.columns)&set(gene_sga_list.iloc[:,0].tolist()))\n",
    "print(len(common_gene_sga))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "1dbcc64b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(976, 1084) (8586, 1084)\n"
     ]
    }
   ],
   "source": [
    "GDSC_sga_df = GDSC_mut_cnv_df[common_gene_sga]\n",
    "TCGA_sga_df = TCGA_mut_cnv_processed_df[common_gene_sga]\n",
    "print(GDSC_sga_df.shape,TCGA_sga_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7a7f5cb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
